{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run \"print_functions.ipynb\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def parsegb(path):\n",
    "    if path[-3:] == \".gz\":\n",
    "        gfl = (gzip.open(path, \"rb\").read()).decode(\"utf-8\") \n",
    "        return list(SeqIO.parse(StringIO(gfl),\"gb\"))\n",
    "    else:\n",
    "        return list(SeqIO.parse(path,\"gb\"))\n",
    "    \n",
    "def extract_name_from_path(path):\n",
    "    return \".\".join(path.split(\"//\")[-1].split(\".\")[:2])\n",
    "    \n",
    "def parsegff(path, return_name = False):\n",
    "    name = extract_name_from_path(path)\n",
    "    id_spec = None\n",
    "    merge_strategy = \"merge\"\n",
    "    if path[-3:] == \".gz\":\n",
    "        gfl = (gzip.open(path, \"rb\").read()).decode(\"utf-8\")\n",
    "        if return_name:\n",
    "            return gffutils.create_db(gfl, ':memory:', from_string = True, id_spec = id_spec,\n",
    "                                      merge_strategy = merge_strategy), name\n",
    "        return gffutils.create_db(gfl, ':memory:', from_string = True, id_spec = id_spec,\n",
    "                                  merge_strategy = merge_strategy)\n",
    "    else:\n",
    "        if return_name:\n",
    "            return gffutils.create_db(path, ':memory:', id_spec = id_spec,\n",
    "                                      merge_strategy = merge_strategy), name\n",
    "        return gffutils.create_db(path, ':memory:', id_spec = id_spec,\n",
    "                                  merge_strategy = merge_strategy)\n",
    "    \n",
    "def parsefasta(path, return_name = False):\n",
    "    name = extract_name_from_path(path)\n",
    "    if path[-3:] == \".gz\":\n",
    "        gfl = (gzip.open(path, \"rb\").read()).decode(\"utf-8\") \n",
    "        if return_name:\n",
    "            return list(SeqIO.parse(StringIO(gfl),\"fasta\")), name\n",
    "        return list(SeqIO.parse(StringIO(gfl),\"fasta\"))\n",
    "    else:\n",
    "        if return_name:\n",
    "            return list(SeqIO.parse(path,\"fasta\"))\n",
    "        return list(SeqIO.parse(path,\"fasta\"))\n",
    "\n",
    "#find strange sequences in gff file\n",
    "def calculate_stats_gff(db):\n",
    "    unusual_ones = []\n",
    "    for elem in list(db.features_of_type(\"gene\")):\n",
    "        cds_cands = (list(db.children(elem)))\n",
    "        cn = 0\n",
    "        for elem in cds_cands:\n",
    "            if elem.featuretype == \"CDS\":\n",
    "                cn += 1\n",
    "        \n",
    "        if cn > 1:\n",
    "            #interesting one\n",
    "            unusual_ones.append((elem, cds_cands))\n",
    "    return unusual_ones, db\n",
    "\n",
    "#placeholders to be implemented\n",
    "def has_gaps(list_of_children):\n",
    "    return None\n",
    "\n",
    "def get_start_codon():\n",
    "    return None\n",
    "\n",
    "#quick path replacement functions\n",
    "def replace_path_gff_to_gbff(path, old_folder, new_folder):\n",
    "    return path.replace(old_folder, new_folder).replace(\".gff\", \".gbff\")\n",
    "def replace_path_gbff_to_gff(path, old_folder, new_folder):\n",
    "    return path.replace(old_folder, new_folder).replace(\".gbff\", \".gff\")\n",
    "\n",
    "#finds feature in fasta list\n",
    "def find_by_id(flist, query):\n",
    "    for f in flist:\n",
    "        if f.id == query:\n",
    "            return f\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_overlapping_genes_one_frame(frame:\"Frame object\"):\n",
    "    frame.genes.sort()\n",
    "    frame.convert_to_computational()\n",
    "    for i in range(1,len(frame.genes)):\n",
    "        if frame.genes[i][0] <= frame.genes[i-1][1]:\n",
    "            print(name, frame[i-1], frame[i])\n",
    "            \n",
    "def select_genes_by_frame(genes, frame, genes_type):\n",
    "    if genes_type != \"computational\":\n",
    "        print(\"WARNING: non-computational type\")\n",
    "    ans = []\n",
    "    for g in genes:\n",
    "        if g[0] % 3 == frame % 3:\n",
    "            #ans.append(g)\n",
    "            ans.append([g[0] - frame % 3, g[1]])\n",
    "    return ans\n",
    "            \n",
    "def computational_split_string(annot:\"Annotation object\"):\n",
    "    annot.convert_to_computational()\n",
    "    ans = []\n",
    "    for frame, ref_frame in ((0,len(annot.seq)),(1,-2),(2,-1)):\n",
    "        frame_genes = select_genes_by_frame(annot.genes, frame, annot.genes_type)\n",
    "        frame_seq = annot.seq[frame:ref_frame]\n",
    "        ans.append(Frame(annot.name, frame_seq, frame, frame_genes, annot.genes_type))\n",
    "\n",
    "    return ans #list of frame obgects\n",
    "\n",
    "def check_same_annot(frame1, frame2):\n",
    "    return (frame1.content in frame2.content) or (frame2.content in frame1.content)\n",
    "\n",
    "def long_wrap(v):\n",
    "    if len(v) > 100:\n",
    "        return tqdm.tqdm(v)\n",
    "    else:\n",
    "        return v\n",
    "\n",
    "def computational_assemble_frames(frames:\"list of frame objects\"):\n",
    "    if len(frames) != 3:\n",
    "        print(\"Warning: frames count != 3\")\n",
    "    #check that frames are from the same annotation\n",
    "    flag = True\n",
    "    for elem1 in frames:\n",
    "        for elem2 in frames:\n",
    "            flag = flag and check_same_annot(elem1, elem2) and (elem1.name == elem2.name)\n",
    "    if not flag:\n",
    "        print(\"Error: frames are not from the same annotation\")\n",
    "        \n",
    "    #now we construct annotation\n",
    "    an_genes = []\n",
    "    an_content = \"\"\n",
    "    an_name = frames[0].name\n",
    "    an_genes_type = frames[0].genes_type\n",
    "    for elem in frames:\n",
    "        if elem.frame == 0:\n",
    "            an_content = elem.content\n",
    "        elem_genes = [[e[0] - elem.frame, e[1] - elem.frame] for e in elem.genes]\n",
    "        an_genes += elem_genes\n",
    "    return Annotation(an_name, an_content, an_genes, an_genes_type)\n",
    "\n",
    "def extract_stats(paths_gff:\"list of paths to .gff files\", paths_fst:\"list of paths to fasta files\",feat_type = \"CDS\", \n",
    "                  drop_fusion_genes = True):\n",
    "    #if two genes have the same start then it is a fusion gene\n",
    "    \n",
    "    full_annotations = []\n",
    "    \n",
    "    for path_gff, path_fst in (zip(paths_gff, paths_fst)):\n",
    "        gff_dat, name = parsegff(path_gff, return_name=True)\n",
    "        fst_dat = parsefasta(path_fst) #list of SeqRecord objectds\n",
    "        #create FullAnnotation object\n",
    "        feat_list = list(gff_dat.features_of_type(feat_type))\n",
    "        full_annotation = FullAnnotation(name, feat_list, fst_dat)\n",
    "        full_annotation.convert_to_computational()\n",
    "        \n",
    "        #for now we want to drop fusion genes to check the simpliest case\n",
    "        if drop_fusion_genes:\n",
    "            full_annotation.drop_fusion_genes()\n",
    "        \n",
    "        full_annotations.append(full_annotation)\n",
    "    return full_annotations\n",
    "\n",
    "def create_train_set(frame:\"Frame object\"):\n",
    "    print(frame.name)\n",
    "    amino_frame = frame.splice_to_amino()\n",
    "    #now we annotate states\n",
    "    possible_states = [\"ncd\", \"cd\", \"start_c\", \"stop_c\", \"model_start\", \"model_end\"]\n",
    "    amino_frame.states = [\"ncd\" for i in range(len(amino_frame.content))]\n",
    "    for gene in amino_frame.genes:\n",
    "        for amin_pos in range(gene[0], gene[1]):\n",
    "            if amino_frame.states[amin_pos] != \"ncd\":\n",
    "                print(\"Train set ruined and it is your fault...\", \"Frame name is\", frame)\n",
    "                print(\"skipping this one\")\n",
    "                return None\n",
    "            if amin_pos == gene[0]:\n",
    "                amino_frame.states[amin_pos] = \"start_c\"\n",
    "            elif amin_pos == gene[1] - 1:\n",
    "                amino_frame.states[amin_pos] = \"stop_c\"\n",
    "            else:\n",
    "                amino_frame.states[amin_pos] = \"cd\"\n",
    "    #amino_frame.convert_to_pomegranate()\n",
    "    return amino_frame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Annotation:\n",
    "    def __init__(self, name, seq, genes, genes_type = \"computational\"):\n",
    "        self.name = name\n",
    "        self.seq = seq\n",
    "        self.genes = genes\n",
    "        #now supports only single stranded annotations\n",
    "        self.strands = \"single\"\n",
    "        #genes type can be either classical or computational. computational -> gene = seq[gene_start,gene_finish]\n",
    "        #classical = human readable genes\n",
    "        self.genes_type = genes_type\n",
    "    def convert_to_computational(self):\n",
    "        if self.genes_type != \"computational\":\n",
    "            #print(self.genes[0])\n",
    "            for gene in self.genes:\n",
    "                gene[0] -= 1 #= [gene[0] - 1, gene[1]]\n",
    "            #print(self.genes[0])\n",
    "        self.genes_type = \"computational\"\n",
    "    \n",
    "    def convert_to_classical(self):\n",
    "        if self.genes_type != \"classical\":\n",
    "            for gene in self.genes:\n",
    "                gene[0] += 1 #= [gene[0] + 1, gene[1]]\n",
    "        self.genes_type = \"classical\"\n",
    "        \n",
    "        \n",
    "    def get_torch_representation(self):\n",
    "        #torch type means that genes are represented as (n,3) array \n",
    "        #with ones in genes on suitable frame and 0 otherwise\n",
    "        #frame is relative to make learning possible\n",
    "        #that means for ATGTGG coords of ones will be ((0,0),(1,1),(2,2),(3,0),(4,1),(5,2))\n",
    "        \n",
    "        new_genes = np.zeros((len(self.seq), 3))\n",
    "        self.convert_to_computational()\n",
    "        for gene in self.genes:\n",
    "            #not sure about that cycle, check after convert\n",
    "            c_frame = 0\n",
    "            for i in range(gene[0], gene[1]):\n",
    "                '''if sum(new_genes[i]) > 0:\n",
    "                    print(new_genes[i], \"old\", c_frame, i)\n",
    "                    new_genes[i][c_frame] = 1\n",
    "                    print(new_genes[i], \"new\")\n",
    "                else:'''\n",
    "                new_genes[i][c_frame] = 1\n",
    "                c_frame += 1\n",
    "                c_frame  %= 3\n",
    "        return (self.seq, new_genes)\n",
    "        \n",
    "    \n",
    "    def append(self, gene):\n",
    "        self.genes.append(gene)\n",
    "        \n",
    "        \n",
    "    def drop_fusion_genes(self, print_dropped = True):\n",
    "        self.genes.sort()\n",
    "        if len(self.genes) == 0:\n",
    "            return\n",
    "        ans_feats = [self.genes[0]]\n",
    "        for i in range(len(self.genes)):\n",
    "            f = self.genes[i]\n",
    "            if (not f[0] == ans_feats[-1][0]) and (not f[1] == ans_feats[-1][1]):\n",
    "                ans_feats.append(f)\n",
    "            elif print_dropped:\n",
    "                print(\"Fusion gene:\", f, self.name)\n",
    "        self.genes = ans_feats\n",
    "\n",
    "\n",
    "\n",
    "class AminoFrame:\n",
    "    def __init__(self, name:\"'name to manually reconstruct all of these'\", \n",
    "                 content:\"cut seq of [YYY] aminoacids\", frame:\"could be either 0, +1 or +2\", \n",
    "                 genes:\"genes in amino coords\", genes_type = \"computational\", states = None):\n",
    "        self.name = name\n",
    "        self.content = content\n",
    "        if not frame in (0,1,2):\n",
    "            print(\"Incorrect frame warning\")\n",
    "        self.frame = frame\n",
    "        self.genes = genes\n",
    "        self.genes_type = genes_type\n",
    "        self.states = states\n",
    "        \n",
    "    #pomegranate format has model_start and model_end bullshit\n",
    "    def convert_to_pomegranate(self):\n",
    "        if self.genes_type != \"pomegranate\":\n",
    "            self.content = [\"AAA\"] + self.content + [\"AAA\"]\n",
    "            self.genes = [[e[0] + 1, e[1] + 1] for e  in self.genes]\n",
    "            self.genes_type = \"pomegranate\"\n",
    "            self.states = [\"None-start\"] + self.states + [\"None-end\"]\n",
    "    def convert_to_computational(self):\n",
    "        if self.genes_type != \"computational\":\n",
    "            self.content = self.content[1:-1]\n",
    "            self.genes = [[e[0] - 1, e[1] - 1] for e  in self.genes]\n",
    "            self.genes_type = \"computational\"\n",
    "            self.states = self.states[1:-1]\n",
    "    \n",
    "    def assemble_to_regular(self):\n",
    "        self.convert_to_computational()\n",
    "        regular_content = \"\".join(self.content)\n",
    "        regular_genes = [[e[0] * 3, e[1] * 3 ] for e in self.genes]\n",
    "        return Frame(self.name, regular_content, self.frame, regular_genes, self.genes_type)\n",
    "    \n",
    "    def create_genes_from_states(self,states):\n",
    "        started = False\n",
    "        current_start = 0\n",
    "        if len(states) != len(self.content):\n",
    "            print(\"Wrong states\")\n",
    "            return False\n",
    "        for i, state in enumerate(states):\n",
    "            if not started:\n",
    "                if state == \"start_c\":\n",
    "                    current_start = i\n",
    "                if state in (\"stop_c\", \"cd\"):\n",
    "                    print(\"Error\")\n",
    "            if started:\n",
    "                if state == \"stop_c\":\n",
    "                    self.genes.append([current_start, i])\n",
    "                    started = False\n",
    "                if state in (\"ncd\", \"start_c\"):\n",
    "                    print(\"Error\")\n",
    "        return True\n",
    "    \n",
    "    def get_start_and_stop_sets(self):\n",
    "        start_s = set()\n",
    "        stop_s = set()\n",
    "        for gene_c in self.genes:\n",
    "            gene_seq = self.content[gene_c[0]: gene_c[1]]\n",
    "            start_s.add(gene_seq[0])\n",
    "            stop_s.add(gene_seq[1])\n",
    "        return start_s, stop_s\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "#frame content does not guaranteed to be div by 3\n",
    "class Frame(Annotation):\n",
    "    def __init__(self, name:\"'name to manually reconstruct all of these'\", \n",
    "                 content:\"cut seq\", frame:\"could be either 0, +1 or +2\", genes, genes_type = \"computational\"):\n",
    "        self.name = name\n",
    "        self.content = content\n",
    "        if not frame in (0,1,2):\n",
    "            print(\"Incorrect frame warning\")\n",
    "        self.frame = frame\n",
    "        self.genes = genes\n",
    "        self.genes_type = genes_type\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return print_to_string(self.name, \", frame:\", self.frame)\n",
    "    \n",
    "    def splice_to_amino(self):\n",
    "        amino_content = []\n",
    "        i = 0\n",
    "        while i + 3 <= len(self.content):\n",
    "            amino_content.append(self.content[i:i+3])\n",
    "            i += 3\n",
    "        amino_genes = [[e[0] // 3, e[1] // 3] for e  in self.genes]\n",
    "        \n",
    "        return AminoFrame(self.name, amino_content, self.frame, amino_genes, self.genes_type)\n",
    "    \n",
    "    \n",
    "class FullAnnotation:\n",
    "    def __init__(self, name, feat_list:\"raw gff dat\", seqs:\"raw fasta dat\", genes_type = \"classical\"):\n",
    "        self.name = name\n",
    "        self.genes_type = genes_type\n",
    "        self.annots = dict(zip([rec.name for rec in seqs],[Annotation(name,rec.seq, [], genes_type) for rec in seqs]))\n",
    "        for f in feat_list:\n",
    "            self.annots[f.seqid].append([f.start,f.stop])\n",
    "    \n",
    "    def convert_to_computational(self):\n",
    "        for annot_key in self.annots:\n",
    "            self.annots[annot_key].convert_to_computational()\n",
    "        self.genes_type = \"computational\"\n",
    "            \n",
    "    def convert_to_classical(self):\n",
    "        for annot_key in self.annots:\n",
    "            self.annots[annot_key].convert_to_classical()      \n",
    "        self.genes_type = \"classical\"\n",
    "        \n",
    "    def drop_fusion_genes(self, print_dropped = True):\n",
    "        for annot_key in self.annots:\n",
    "            self.annots[annot_key].drop_fusion_genes(print_dropped=print_dropped)      \n",
    "    \n",
    "    def __repr__(self):\n",
    "        return print_to_string(self.name, \"genes_type:\", self.genes_type, \"with\", len(self.annots), \"annotation(s)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'abc'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = \"abcdefg\"\n",
    "a[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "finalized": {
   "timestamp": 1629911397227,
   "trusted": true
  },
  "kernelspec": {
   "display_name": "torch",
   "language": "python",
   "name": "torch"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
